subsys/pm/state.c

```diff
@@ -40,12 +40,20 @@ BUILD_ASSERT(DT_NODE_EXISTS(DT_PATH(cpus)),
 DT_FOREACH_CHILD(DT_PATH(cpus), CHECK_POWER_STATES_CONSISTENCY)
 
 #define DEFINE_CPU_STATES(n) \
-	static const struct pm_state_info pmstates_##n[] \
+	static const struct pm_state_info pmstates_cpu##n[] \
 		= PM_STATE_INFO_LIST_FROM_DT_CPU(n);
-#define CPU_STATE_REF(n) pmstates_##n
+#define CPU_STATE_REF(n) pmstates_cpu##n
 
 DT_FOREACH_CHILD(DT_PATH(cpus), DEFINE_CPU_STATES);
 
+#if CONFIG_MP_NUM_CPUS > 1 && DT_NODE_EXISTS(DT_PATH(cpus, power_states))
+/** General power-states list on SMP platforms, as individual per-CPU states may vary */
+static const struct pm_state_info pmstates_all[] = {
```

Comment:
This guard condition turns out to be somewhat problematic:
In my use case #59200, a state can be defined in `power-states` and `reinit-power-states`, but be missing from `cpu-power-states`. This'll cause a crash when we register a notifier for that missing cpu power state, as we'd try to reference a notifier which doesn't exist.

In summary, we kind of need both lists and shouldn't exclude pmstates_all on single core platforms :/

We

Reply:
To be sure IIUC, `stop2` is `reinit-power-states` on usart1, as this is a configuration and not in `cpu-power-states` because this is your application configuration. Am I correct ?

Reply:
Yes, this is just a corner case, but I think we should cover it as the resulting errors aren't entirely obvious. What do others think though?

Reply:
Any thoughts @ceolin & @gmarull?

Reply:
Turns out [I'm an idiot](https://github.com/zephyrproject-rtos/zephyr/pull/60998/files#diff-d4e51fe3d6c0c3c82cef6a93687ea6246bc506c89f069f071138be1d497459d6R143-R145).